bevy_lint/src/utils/hir_parse.rs

```diff
@@ -0,0 +1,44 @@
+//! Utility functions for parsing HIR types.
+
+use rustc_hir::{GenericArg, Node, QPath, Ty, TyKind};
+use rustc_lint::LateContext;
+
+/// Returns the list of types inside a tuple type.
+///
+/// If the type is not a tuple, returns a list containing the type itself.
+pub(crate) fn detuple(ty: Ty<'_>) -> Vec<Ty<'_>> {
+    if let TyKind::Tup(items) = ty.peel_refs().kind {
+        items.to_vec()
+    } else {
+        vec![ty]
+    }
+}
+
+/// Gets the [`Ty`] for a generic argument at the specified index.
+///
+/// If the generic argument is not a type, returns `None`.
+pub(crate) fn generic_type_at<'tcx>(
+    cx: &LateContext<'tcx>,
+    hir_ty: &'tcx Ty<'tcx>,
+    index: usize,
+) -> Option<&'tcx Ty<'tcx>> {
+    let generic_arg = generic_at(hir_ty, index)?;
+    let generic_node = cx.tcx.hir_node(generic_arg.hir_id());
+
+    if let Node::Ty(ty) = generic_node {
+        Some(ty)
+    } else {
+        None
+    }
+}
+/// Returns the [`GenericArg`] at the given index.
+pub(crate) fn generic_at<'hir>(
+    hir_ty: &'hir Ty<'hir>,
+    index: usize,
+) -> Option<&'hir GenericArg<'hir>> {
+    let TyKind::Path(QPath::Resolved(_, path)) = hir_ty.kind else {
+        return None;
+    };
+
+    path.segments.last()?.args().args.get(index)
```

Comment:
On second thought, this might break in cases like:

```rust
trait Queryable {
    type Query<'w, 's, D: QueryData + 'static, F: QueryFilter + 'static>;
}

struct MyQueryable<T>(T);

impl<T> Queryable for MyQueryable<T> {
    type Query<'w, 's, D: QueryData + 'static, F: QueryFilter + 'static> = Query<'w, 's, D, F>;
}

fn system2<'w, 's>(query: <MyQueryable<f32> as Queryable>::Query<'w, 's, &'static Foo, ()>) {}
```

Although, to be fair, I just tested this and we don't currently catch associated types like this.

Reply:
Maybe we save adding this functionality for the future when it's needed? It might at that point even make sense to be a separate function depending on the edge cases.

Reply:
> Maybe we save adding this functionality for the future when it's needed?

I'm fine with that, but would you mind adding a comment to the function docs do users know that it only indexes the generics in the final path segment? (Non-blocking, though)